src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -335,8 +365,7 @@ fn rustc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
                 if !src_dir.ends_with("deps") {
                     continue
                 }
-                let dst = src_dir.parent().unwrap()
-                                 .join(src.file_name().unwrap());
+                let dst = src_dir.parent().unwrap().join(bin_name);
```

Comment:
Might this code below (hard link when doing move_outputs_up) also be susceptible to the same bug? Perhaps not today, since we should only be moving up a top level package, but it seems like a fragility that might cause us to inadvertently add a bug in the future.

One possibility (striving for simplicity over performance) is to only copy and never link.